src/lazy/encoder/writer.rs

```diff
@@ -0,0 +1,501 @@
+use std::io::Write;
+
+use delegate::delegate;
+use ice_code::ice as cold_path;
+
+use crate::constants::v1_0::system_symbol_ids;
+use crate::lazy::encoder::annotation_seq::AnnotationSeq;
+use crate::lazy::encoder::value_writer::internal::{FieldEncoder, MakeValueWriter};
+use crate::lazy::encoder::value_writer::{
+    AnnotatableWriter, EExpWriter, SequenceWriter, StructWriter, ValueWriter,
+};
+use crate::lazy::encoder::{LazyEncoder, LazyRawWriter, SymbolCreationPolicy};
+use crate::lazy::text::raw::v1_1::reader::MacroIdRef;
+use crate::raw_symbol_token_ref::AsRawSymbolTokenRef;
+use crate::result::IonFailure;
+use crate::{
+    Decimal, Element, ElementWriter, Int, IonResult, IonType, RawSymbolTokenRef, Symbol, SymbolId,
+    SymbolTable, Timestamp, Value, WriteConfig,
+};
+
+pub(crate) struct EncodingContext {
+    symbol_table: SymbolTable,
+    num_pending_symbols: usize,
+    symbol_creation_policy: SymbolCreationPolicy,
+    supports_text_tokens: bool,
+}
+
+impl EncodingContext {
+    pub fn new(
+        symbol_table: SymbolTable,
+        symbol_creation_policy: SymbolCreationPolicy,
+        supports_text_tokens: bool,
+    ) -> Self {
+        Self {
+            symbol_table,
+            num_pending_symbols: 0,
+            symbol_creation_policy,
+            supports_text_tokens,
+        }
+    }
+}
+
+/// An Ion writer that maintains a symbol table and creates new entries as needed.
+pub struct ApplicationWriter<E: LazyEncoder, Output: Write> {
+    encoding_context: EncodingContext,
+    data_writer: E::Writer<Vec<u8>>,
+    directive_writer: E::Writer<Vec<u8>>,
+    output: Output,
+}
+
+impl<E: LazyEncoder, Output: Write> ApplicationWriter<E, Output> {
+    /// Constructs a writer for the requested encoding using its default configuration.
+    pub fn new(output: Output) -> IonResult<Self> {
+        Self::with_config(E::default_write_config(), output)
+    }
+
+    /// Constructs a writer for the requested encoding using the provided configuration.
+    pub fn with_config(config: WriteConfig<E>, output: Output) -> IonResult<Self> {
+        let directive_writer = E::Writer::build(config.clone(), vec![])?;
+        let mut data_writer = E::Writer::build(config, vec![])?;
+        // Erase the IVM that's created by default
+        data_writer.output_mut().clear();
```

Comment:
Could you put the initial version into the config, so that the writer always starts with the correct IVM instead of having to clear a default IVM?

Reply:
Now that there's an application-level writer that can automate calling a `write_ivm()` method, I think we could remove the automatic functionality from the raw-level writers.

There's a minor design question around what to do for the 1.0 text writer. When the stream starts, it could leave off the `$ion_1_0`. However, if the user or application writer is calling `write_ivm()` to reset the encoding context, it could not leave it out. We could either have it _always_ include a `$ion_1_0` (unnecessary but pretty harmless) or we could pass an `implicit_ivm_permitted: bool` that other writers could just ignore.

Reply:
I opened #746 to track this.